eZ/Publish/Core/MVC/Symfony/Templating/Twig/Extension/RoutingExtension.php

```diff
@@ -43,8 +69,75 @@ public function getName()
      *
      * @return \eZ\Publish\Core\MVC\Symfony\Routing\RouteReference
      */
-    public function getRouteReference($resource = null, $params = [])
+    public function getRouteReference($resource = null, $params = []): RouteReference
     {
         return $this->routeReferenceGenerator->generate($resource, $params);
     }
+
+    public function getPath($name, array $parameters = [], bool $relative = false): string
+    {
+        $referenceType = $relative ? UrlGeneratorInterface::RELATIVE_PATH : UrlGeneratorInterface::ABSOLUTE_PATH;
+        if (is_object($name)) {
+            return $this->generateUrlForObject($name, $parameters, $referenceType);
+        }
+
+        return $this->urlGenerator->generate($name, $parameters, $referenceType);
+    }
+
+    public function getUrl($name, array $parameters = [], bool $schemeRelative = false): string
+    {
+        $referenceType = $schemeRelative ? UrlGeneratorInterface::NETWORK_PATH : UrlGeneratorInterface::ABSOLUTE_URL;
+        if (is_object($name)) {
+            return $this->generateUrlForObject($name, $parameters, $referenceType);
+        }
+
+        return $this->urlGenerator->generate($name, $parameters, $referenceType);
+    }
+
+    private function generateUrlForObject(object $object, array $parameters, int $referenceType): string
+    {
+        if ($object instanceof Location) {
+            $routeName = UrlAliasRouter::URL_ALIAS_ROUTE_NAME;
+            $parameters += [
+                'locationId' => $object->id,
+            ];
+        } elseif ($object instanceof Content || $object instanceof ContentInfo) {
+            $routeName = UrlAliasRouter::URL_ALIAS_ROUTE_NAME;
+            $parameters += [
+                'contentId' => $object->id,
+            ];
+        } elseif ($object instanceof RouteReference) {
+            $routeName = $object->getRoute();
+            $parameters += $object->getParams();
+        } else {
+            $routeName = '';
+            $parameters += [
+                RouteObjectInterface::ROUTE_OBJECT => $object,
+            ];
+        }
```

Comment:
This is not very SOLID, you should have separate strategy handler for each type of object and use visitor pattern. Strategy should call `$this->urlGenerator->generate`. Here only a fallback should be handled.